lib/semantics/check-do-concurrent.cc

```diff
@@ -339,7 +339,8 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
       explicit CollectSymbols(int) {}
       void Handle(const Symbol *symbol) { result().push_back(symbol); }
     };
-    return evaluate::Visitor<CollectSymbols>{0}.Traverse(*expression.typedExpr);
+    return evaluate::Visitor<CollectSymbols>{0}.Traverse(
+        expression.typedExpr->v);
```

Comment:
I think it is clearer if we avoid `typedExpr` and use `GetExpr()` instead:
```
  if (const auto *expr{GetExpr(expression)}) {
    ...
    return evaluate::Visitor<CollectSymbols>{0}.Traverse(*expr);
  } else ...
```

Also, we should just get rid of the "else" part. If there isn't an analyzed expression we should just skip it. We don't need an alternate implementation.